CIP-XXXX/CIP-XXXX.md

```diff
@@ -0,0 +1,558 @@
+distribution is equally probable, we need $\Theta(8 \cdot n)$ time in the
+average case. While it is
+impossible to do both of these operations in sub-linear time in general, the
+large constant factors this imposes (as well as the overhead of looping over
+_bit_ indexes) is a cost we can ill afford on-chain, especially if the goal is
+to use these operations as 'building blocks' for something like a data
+structure.
+
+### Case 2: hashing
+
+In our [previously-described][hashing] case, we stated what operations we would
+need for the Argon2 family of hashes specifically. However, Argon2 has a
+specific advantage in that the number of operations it requires are both
+relatively few, and the most complex of which (BLAKE2b hashing) already exists
+in Plutus Core as a primitive. However, other hash functions (and indeed, many
+other cryptographic primitives) rely on two other important instructions: bit
+shifts and bit rotations. As an example, consider SHA512, which is an important
+component in several cryptographic protocols (including Ed25519 signature
+verification): its implementation requires both shifts and rotations to work. 
+
+Like with Case 1, we can theoretically simulate both rotations and shifts using
+a combination of bit reads and bit writes to an empty `BuiltinByteString`.
+However, the cost of this is extreme: we would need to produce a list of
+index-value pairs of length equal to the Hamming weight of the input, only to
+then immediately discard it! To put this into some perspective, for an 8-byte
+input, performing a rotation involves allocating an expected 32 index-value
+pairs, using _significantly_ more memory than the result. On-chain, we can't
+really afford this cost, especially in an operation intended to be used as part
+of larger constructions (as would be necessary here).
+
+## Specification
+
+We describe the proposed operations in several stages. First, we give an
+overview of the proposed operations' signatures and costings; second, we
+describe the semantics of each proposed operation in detail, as well as some
+examples. Lastly, we provide laws that any implementation of the proposed
+operations should obey.
+
+Throughout, we make use of the [bit indexing scheme][bit-indexing-scheme]
+described in a CIP-122. We also re-use the notation $x[i]$ to refer to the value
+of at bit index $i$ of $x$, and the notation $x\\{i\\}$ to refer to the byte at
+byte index $i$ of $x$: both are specified in CIP-122.
+
+### Operation semantics
+
+Our proposed operations will have the following signatures:
+
+* ``bitwiseShift :: BuiltinByteString -> BuiltinInteger -> BuiltinByteString``
+* ``bitwiseRotate :: BuiltinByteString -> BuiltinInteger -> BuiltinByteString``
+* ``countSetBits :: BuiltinByteString -> BuiltinInteger``
+* ``findFirstSetBit :: BuiltinByteString -> BuiltinInteger``
+
+We assume the following costing, for both memory and execution time:
+
+| Operation | Execution time cost | Memory cost |
```

Comment:
For some reason this table isn't rendering properly, at least on GitHub.

Reply:
That was me failing at GFM table syntax. Fixed.